### DISCUSSION TITLE
How to add to dp.Report for dynamic reporting

### DISCUSSION THREAD

**Discussion by UltimateLaForsch** [datapane/datapane#71] [Help]
Hi,
I' just trying out DataPane for a reporting project. Now, I would like to use it for a larger project with dynamic content. E.g. I got datarows from database. These rows, I would like to filter for a certain criterion and then loop the resulting subset by another attribute, adding a table and two graphs for each attribute. This subset can be of different size depending on monthly activity. Therefore, it is not known in advance how many graphs it would need to create. Is it possible to add to db.report (e.g. db.Table and db.Plot objects) during the loop process? So, I can add the complete db.Report with all content to report...
e.g. 
```python
for x in db_rows.index:
          fig = ... some code to generate a dataseries for each loop
          text = db.Text(some index-related text)
          table =  db.Table(a index-related table)
          plot = db.Plot(fig)
     report_attributes.append/update/.... (text, table, plot)
report = db.Report(report_attributes)
report.save(...)
```

Thx

**Comment by mands** [comment]
Yes, Datapane supports dynamic reports by design. The report object takes either a list of the elements or a list blocks - which you can edit/modify as needed. One easy way to do it is to create a blocks list of the top-level elements, edit that as needed, then create your report with that blocks list, e.g.

```python
blocks = []
for x in db_rows.index:
          fig = ... some code to generate a dataseries for each loop
          text = db.Text(some index-related text)
          table =  db.Table(a index-related table)
          plot = db.Plot(fig)
          blocks.append([text, table, plot])
report = db.Report(blocks=blocks)
report.save(...)
```

You can even build up helper functions that will return the sections for you, wrapped up as a list of blocks, e.g.

```python
def per_row_section(x) -> List[dp.Block]:
          fig = ... some code to generate a dataseries for each loop
          text = db.Text(some index-related text)
          table =  db.Table(a index-related table)
          plot = db.Plot(fig)
          return [f"Section {x}", text, table, plot]

report = db.Report(blocks=[y for x in db_row.index for y in per_row_section(x)])
```

There is also features such as `dp.Group` to collect together subsections that can be dynamically added to a the report.
